File: docx_parser_converter/docx_parsers/helpers/common_helpers.py
from typing import Optional
from xml.etree.ElementTree import Element

NAMESPACE_URI = "http://schemas.openxmlformats.org/wordprocessingml/2006/main"
NAMESPACE = {'w': NAMESPACE_URI}

def extract_element(parent: Optional[Element], path: str) -> Optional[Element]:
    """
    Extracts an XML element from the parent element using the given path.
    
    Args:
        parent (Optional[Element]): The parent XML element.
        path (str): The XPath to the desired child element.

    Returns:
        Optional[Element]: The extracted XML element, or None if not found or if parent is None.

    Example:
        The following is an example of extracting a paragraph properties element 
        from a paragraph element in a document.xml file:

        .. code-block:: xml

            <w:p>
                <w:pPr>
                    <!-- Paragraph properties here -->
                </w:pPr>
            </w:p>

        Usage:
            pPr = extract_element(paragraph_element, ".//w:pPr")
    """
    if parent is None:
        return None
    return parent.find(path, namespaces=NAMESPACE)

def extract_attribute(element: Optional[Element], attr: str) -> Optional[str]:
    """
    Extracts an attribute value from an XML element.
    
    Args:
        element (Optional[Element]): The XML element.
        attr (str): The attribute name.

    Returns:
        Optional[str]: The attribute value, or None if not found.

    Example:
        The following is an example of extracting the 'val' attribute from a style 
        element in a document.xml file:

        .. code-block:: xml

            <w:pStyle w:val="Heading1"/>

        Usage:
            style_val = extract_attribute(style_element, 'val')
    """
    if element is not None:
        return element.get(f'{{{NAMESPACE_URI}}}{attr}')
    return None

def safe_int(value: Optional[str]) -> Optional[int]:
    """
    Converts a string value to an integer safely.
    
    Args:
        value (Optional[str]): The string value to convert.

    Returns:
        Optional[int]: The integer value, or None if the input is None.

    Example:
        The following is an example of safely converting a string to an integer:

        .. code-block:: python

            int_value = safe_int("123")  # Returns 123
            int_value = safe_int(None)  # Returns None
    """
    return int(value) if value is not None else None

def extract_boolean_attribute(element: Optional[Element]) -> Optional[bool]:
    """
    Extracts a boolean attribute from an XML element.
    
    Args:
        element (Optional[Element]): The XML element.

    Returns:
        Optional[bool]: True if the element is present and its 'val' attribute is not 'false' or '0',
                        False if its 'val' attribute is 'false' or '0',
                        None if the element is not present.

    Example:
        The following is an example of extracting a boolean attribute from an element:

        .. code-block:: xml

            <w:b w:val="true"/>

        Usage:
            bold = extract_boolean_attribute(bold_element)  # Returns True if w:val is not 'false' or '0'
    """
    if element is not None:
        val = element.get(f'{{{NAMESPACE_URI}}}val')
        if val is not None:
            return val.lower() not in ["false", "0"]
        return True
    return None

File: docx_parser_converter_ts/src/docx_parsers/helpers/commonHelpers.ts
export const NAMESPACE_URI = "http://schemas.openxmlformats.org/wordprocessingml/2006/main";
export const NAMESPACE = { w: NAMESPACE_URI };

/**
 * Extracts an XML element from the parent element using the given path.
 * 
 * @param parent - The parent XML element.
 * @param path - The XPath to the desired child element.
 * @returns The extracted XML element, or null if not found or if parent is null.
 * 
 * @example
 * The following is an example of extracting a paragraph properties element 
 * from a paragraph element in a document.xml file:
 * ```xml
 * <w:p>
 *     <w:pPr>
 *         <!-- Paragraph properties here -->
 *     </w:pPr>
 * </w:p>
 * ```
 * 
 * Usage:
 * ```typescript
 * const pPr = extractElement(paragraphElement, ".//w:pPr");
 * ```
 */
export function extractElement(parent: Element | null, path: string): Element | null {
  if (parent === null) {
    return null;
  }

  // Create a namespace resolver function
  const nsResolver = (prefix: string | null): string | null => {
    if (prefix === 'w') {
      return NAMESPACE_URI;
    }
    return null;
  };

  // Use XPath evaluation to find the element
  const result = document.evaluate(
    path,
    parent,
    nsResolver,
    XPathResult.FIRST_ORDERED_NODE_TYPE,
    null
  );

  return result.singleNodeValue as Element | null;
}

/**
 * Extracts an attribute value from an XML element.
 * 
 * @param element - The XML element.
 * @param attr - The attribute name.
 * @returns The attribute value, or null if not found.
 * 
 * @example
 * The following is an example of extracting the 'val' attribute from a style 
 * element in a document.xml file:
 * ```xml
 * <w:pStyle w:val="Heading1"/>
 * ```
 * 
 * Usage:
 * ```typescript
 * const styleVal = extractAttribute(styleElement, 'val');
 * ```
 */
export function extractAttribute(element: Element | null, attr: string): string | null {
  if (element !== null) {
    return element.getAttributeNS(NAMESPACE_URI, attr);
  }
  return null;
}

/**
 * Converts a string value to a number safely.
 * 
 * @param value - The string value to convert.
 * @returns The number value, or null if the input is null or invalid.
 * 
 * @example
 * The following is an example of safely converting a string to a number:
 * ```typescript
 * const intValue = safeInt("123");  // Returns 123
 * const intValue = safeInt(null);   // Returns null
 * const intValue = safeInt("abc");  // Returns null
 * ```
 */
export function safeInt(value: string | null): number | null {
  if (value === null) {
    return null;
  }
  
  const parsed = parseInt(value, 10);
  return isNaN(parsed) ? null : parsed;
}

/**
 * Converts a string value to a floating point number safely.
 * 
 * @param value - The string value to convert.
 * @returns The number value, or null if the input is null or invalid.
 * 
 * @example
 * ```typescript
 * const floatValue = safeFloat("123.45");  // Returns 123.45
 * const floatValue = safeFloat(null);      // Returns null
 * const floatValue = safeFloat("abc");     // Returns null
 * ```
 */
export function safeFloat(value: string | null): number | null {
  if (value === null) {
    return null;
  }
  
  const parsed = parseFloat(value);
  return isNaN(parsed) ? null : parsed;
}

/**
 * Extracts a boolean attribute from an XML element.
 * 
 * @param element - The XML element.
 * @returns True if the element is present and its 'val' attribute is not 'false' or '0',
 *          False if its 'val' attribute is 'false' or '0',
 *          null if the element is not present.
 * 
 * @example
 * The following is an example of extracting a boolean attribute from an element:
 * ```xml
 * <w:b w:val="true"/>
 * ```
 * 
 * Usage:
 * ```typescript
 * const bold = extractBooleanAttribute(boldElement);  // Returns true if w:val is not 'false' or '0'
 * ```
 */
export function extractBooleanAttribute(element: Element | null): boolean | null {
  if (element !== null) {
    const val = element.getAttributeNS(NAMESPACE_URI, 'val');
    if (val !== null) {
      return val.toLowerCase() !== "false" && val !== "0";
    }
    return true;
  }
  return null;
}

/**
 * Extracts all matching elements from the parent element using the given XPath.
 * 
 * @param parent - The parent XML element.
 * @param path - The XPath to the desired child elements.
 * @returns Array of extracted XML elements, or empty array if none found or if parent is null.
 * 
 * @example
 * ```typescript
 * const runs = extractElements(paragraphElement, ".//w:r");
 * ```
 */
export function extractElements(parent: Element | null, path: string): Element[] {
  if (parent === null) {
    return [];
  }

  const nsResolver = (prefix: string | null): string | null => {
    if (prefix === 'w') {
      return NAMESPACE_URI;
    }
    return null;
  };

  const result = document.evaluate(
    path,
    parent,
    nsResolver,
    XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,
    null
  );

  const elements: Element[] = [];
  for (let i = 0; i < result.snapshotLength; i++) {
    const node = result.snapshotItem(i) as Element;
    if (node) {
      elements.push(node);
    }
  }

  return elements;
} 